### PR TITLE
Iteration-1

### PR DESCRIPTION
**Pull request by liene-rg**
Code for iteration-1 of Game of Life.

### CODE REVIEW

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: Use summary comments. You can write  '///' line before parameter/method and visual studio will generate summary comment template.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: these properties can be private.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: All this does is copy nextGen into currentGen ? You can use .Clone() for this purpose.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs:32]: Have to get used to commenting your code.
Summary comments for methods are must as these comments are displayed by IntelliSense when calling these methods.

You can write '///' line before method and visual studio will generate comment template.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: This if can if can be combined with if above.

File: Game_Graphic.cs
Comment (EdgarsBondars) [inline Game_Graphic.cs]: Code style has to be consistent everywhere if you use underscore as some kind of separator use it everywhere.
But I would suggest you stick to PascalCase without using underscores.

File: Program.cs
Comment (EdgarsBondars) [inline Program.cs]: You can get rid of temp variable and read user input in while condition just like you did on line 23.

File: Game_Graphic.cs
Comment (EdgarsBondars) [inline Game_Graphic.cs]: Property names should be self explanatory. When creating variable names try to avoid using abbreviations or shortening words.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: You can move these to constructor and delete warning pragma's.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: ( x + i < 0 || x + i >= **_x_** )
I think it was supposed to be width/height check here instead of x ?  :)

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: if (x + i == x && y + **_-j_** == y)
any reason why j should be negative :) ?

File: Program.cs
Comment (EdgarsBondars) [inline Program.cs]: Any specific reason why board has to be over 90 cells in width and 40 in height ?

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Typo in the code, typed > 2 instead of < 2 
if (currentGen[x, y] == 1 && liveNeighbours > 2) was meant less than 2

File: Game_Graphic.cs
Comment (liene-rg) [reply Game_Graphic.cs]: Property gol renamed to gameOfLife.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Moved to constructor. Warnings deleted.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Correct, was supposed to be Height and Width. Code updated.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Typo mistake, value of j should not be negative. Code updated.

File: Game_Graphic.cs
Comment (GuntarsD) [reply Game_Graphic.cs]: Better stick to PascalCase.

File: Game_Graphic.cs
Comment (GuntarsD) [inline Game_Graphic.cs]: Variable names should be made with full names except few widely used abbreviations.
It could be something like:
gameData = gameObject;

File: Program.cs
Comment (GuntarsD) [inline Program.cs]: Here and in other places comment starts with upper case letter and end with dot.
Comments inside methods should be placed on line preceding code to be commented and line before comment should be empty.

File: Program.cs
Comment (GuntarsD) [inline Program.cs]: Empty lines should be removed.
There should be empty line after closed code block, but not before other closing }.

File: Game.cs
Comment (GuntarsD) [inline Game.cs]: Use full words for names.
It should be:
nextGeneration

File: Game.cs
Comment (GuntarsD) [inline Game.cs]: You should check the logic.
This test is wrong.
In case of three neighbors cell will be alive in text generation.
That is true both for alive or dead cell.

File: Game_Graphic.cs
Comment (GuntarsD) [inline Game_Graphic.cs]: It is not good to make sequential concatenations(adding up string many times). In C# string is immutable so adding new part creates next copy of string. That is slow and consumes memory. When that is actually needed then StringBuilder class should be used.

In this particular case you can avoid using buffer at all, because you can print symbols as they are processed.
Use Console.Write("#"); instead of buffer += "#";
And when given line has been printed, that is outside for block, 
use Console.WriteLine(): to continue printing on new line.

File: Program.cs
Comment (GuntarsD) [inline Program.cs]: You have done well creating separate class for logic related to game graphics, but you can use the same correct approach and create separate class for interaction with user.
Code related to retrieving data from user could be put into this class. That would give you opportunity create methods which could be used repeatedly.
Here we have code which retrieves height of the board, but that is almost the same code as following code which retrieves width of board. 
Create method which receives text prompt to be displayed for user, processes user's input and returns numeric value.
 That class could be made static so usage would be convenient.

File: Game.cs
Comment (liene-rg) [reply Game.cs:32]: Summary comments using /// applied.

File: Program.cs
Comment (liene-rg) [reply Program.cs]: The board did not update (move) if the cell count was low, issue fixed.

File: Game_Graphic.cs
Comment (liene-rg) [reply Game_Graphic.cs]: Class name changed to GameGraphic.
gol changed to gameOfLife.
gameObj changed to gameObject.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: currentGen updated to currentGeneration.
nextGen updated to nextGeneration.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: There was a typo in the code. It was meant to be (liveNeighbours < 2) instead of (liveNeighbours > 2). Code updated.

File: Game_Graphic.cs
Comment (liene-rg) [reply Game_Graphic.cs]: Variable buffer removed, code updated.

if (gameOfLife.currentGeneration[i, j] == 1)
                        Console.Write("*");
                    else
                        Console.Write(" ");
                }
            Console.WriteLine();


** Thank you for the explanation and advice.

File: Game.cs
Comment (GuntarsD) [reply Game.cs]: That looks wrong as well.
Try this
image

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Summary comments created.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Code updated.
public int[,] currentGeneration { get; private set; }
private int[,] nextGeneration;

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Switch statement applied.

File: Program.cs
Comment (liene-rg) [reply Program.cs]: Comments inside methods updated.

File: Program.cs
Comment (liene-rg) [reply Program.cs]: Empty lines in the code removed.

File: Program.cs
Comment (liene-rg) [reply Program.cs]: Class UserInput created with 2 static methods UserInputHeight() and UserInputWidth() that process user input and return integer value.

**Comment by liene-rg** [comment]
Reviewed and replied to comments. Updated code pushed to branch iteration1.

File: Program.cs
Comment (GuntarsD) [reply Program.cs]: Good, but can be improved. Right now you have two methods which are performing almost the same functionality. It even could be that you copied one method to create other. Best approach would be to make method like:
public static int GetUserInput(string promptMessage)
So you provide message describing what data should be inserted. Then you could call it:
int numberOfRows = GetUserInput("Enter the height of the board.");
or
int numberOfColumns = GetUserInput("Enter the width of the board.");

Additionally you could make:
public static int GetValidatedUserInput(string promptMessage, int minValue, int maxValue)
To have validation of value at the same place.
This second method could use first one to forward message and get value. Then perform only number validation.

File: GameGraphic.cs
Comment (GuntarsD) [inline GameGraphic.cs:31]: Having one code statement after if or for is valid, but may cause errors, so in professional code there always should be block containing one or many statements. Like:
if (gameOfLife.currentGeneration[i, j] == 1)
{
     Console.Write("*");
 }
 
 Please change where appropriate.

File: UserInput.cs
Comment (GuntarsD) [inline UserInput.cs]: Texts and number literals shouldn't be used directly in code. The values should be placed in separate resource file or placed at start of a class assigned to const variables. Then they should be used in place of texts in code.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs:14]: Random empty line between property declaration.
Code style has to be consistent everywhere you either put put empty line after property or you don't. Same thing with empty line between methods some places you have it in some you don't.

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: Don't forget to write parameter and return comments too :)

File: Game.cs
Comment (EdgarsBondars) [inline Game.cs]: Can't you just do currentGeneration = nextGeneration here ?

File: GameGraphic.cs
Comment (EdgarsBondars) [inline GameGraphic.cs:25]: For me after running your game it's printing everything in one line you have to add curly brackets after for statement and make sure it adds new line after a line is written.

File: Program.cs
Comment (EdgarsBondars) [inline Program.cs]: All the contents of this Main class could go to GameGraphic or some new class some kind of GameHandler. Main class should be just few lines where we initialize game and run it.

File: Program.cs
Comment (EdgarsBondars) [inline Program.cs]: Program should not inherit anything.
Make UseInput static class and make calls like
```
UserInput.UserInputHeight()
````
because seeing just 
```
UserInputHeight()
````
is kinda magic method that comes from nowhere.

File: UserInput.cs
Comment (EdgarsBondars) [inline UserInput.cs]: 1. Naming this variable temp is confusing could be something like userInput.
2. You don't need this temp variable at all you can get user input straight away in while condition like
```
while (!Int32.TryParse(Console.Readline(), out x))
````

**Review by EdgarsBondars — CHANGES_REQUESTED** [review]

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Removed the method as it was not really needed, instead used where necessary currentGeneration = nextGeneration;

File: Program.cs
Comment (liene-rg) [reply Program.cs]: UserInput class updated. public static int GetUserInput(string promptMessage) method that receives user input and validates it.

File: GameGraphic.cs
Comment (liene-rg) [reply GameGraphic.cs:31]: { } after if statements added.

File: UserInput.cs
Comment (liene-rg) [reply UserInput.cs]: Variables created and inserted in the code.

File: Game.cs
Comment (liene-rg) [reply Game.cs:14]: Empty lines removed, went through the code to check for empty lines and inconsistencies in formatting.
Empty line after } , for instance, after the closing } for a method.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: Comments added.

File: Game.cs
Comment (liene-rg) [reply Game.cs]: TransferNextGeneration(); method removed as it was no longer needed, code updated with currentGeneration = nextGeneration .

File: GameGraphic.cs
Comment (liene-rg) [reply GameGraphic.cs:25]: The necessary {} added.
Thanks for noticing the mistake, I was not able to figure out how to print the game nicely. Beginners mistake :)

File: Program.cs
Comment (liene-rg) [reply Program.cs]: GameHandler class created with a method public static void RunGame(Game gameObject, GameGraphic gameGraphic).
Reduced the lines of code in Main.

File: Program.cs
Comment (liene-rg) [reply Program.cs]: Inherit statement removed and methods called properly.

File: UserInput.cs
Comment (liene-rg) [reply UserInput.cs]: temp variable removed and UserInput method updated.

File: UserInput.cs
Comment (GuntarsD) [inline UserInput.cs:23]: Empty line before comment, after closing block( before: return validUserInput; ) and put comments for parameter and return value between opening and closing tag:
Otherwise small, neat method which fits well into screen and is easy to understand.

File: UserInput.cs
Comment (EdgarsBondars) [inline UserInput.cs]: Parameter and return comments should be written between corresponding tags
like
```
        /// <summary>
        /// Gets the value of either Height or Width from the user.
        /// </summary>
        /// <param name="promptMessage">Message that will be displayed to user.</param>
        /// <returns>Returns an integer value from user input.</returns>
````
Check other places too.

File: GameGraphic.cs
Comment (GuntarsD) [inline GameGraphic.cs]: IF block is not needed. Can be replaced by assignment.

File: UserInput.cs
Comment (liene-rg) [reply UserInput.cs]: Comments written between tags.

File: UserInput.cs
Comment (liene-rg) [reply UserInput.cs:23]: Method formatted.

File: GameGraphic.cs
Comment (liene-rg) [reply GameGraphic.cs]: If block removed.

**Comment by liene-rg** [comment]
Code updated related to comments. Waiting for approval for iteration1.

**Review by GuntarsD — APPROVED** [review]

**Review by EdgarsBondars — APPROVED** [review]

**Pull request merged**
